libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +915,75 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, false);
+}
+
+string YulUtilFunctions::readFromCalldata(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, true);
+}
+
+string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromCalldata)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
```

Comment:
Please add an assertion: `if (_fromCalldata) solAssert(!_type.isDynamicallyEncoded(), "")`

Reply:
But I think we should really use the ABI decoding functions for calldata and not reimplement everything here.

Reply:
Basically having two separate functions? Or calling the ABI things from this one?

Reply:
Whatever makes more sense from an architectural standpoint.

Reply:
But you haven't really done calldata anyway, this is just a reminder.